### PR TITLE
[JENKINS-48227] Use "Files.createTempDirectory" to create temp directory

### PR DESCRIPTION
**Pull request by larrysingleton**
Jira contains link to sonar squid number as well as link to personal sonar cloud analysis report relating to this issue

### CODE REVIEW

File: core/src/main/java/hudson/FilePath.java [hunk 1]
```diff
@@ -1400,17 +1402,15 @@ public String invoke(File dir, VirtualChannel channel) throws IOException {
...
     public FilePath createTempDir(final String prefix, final String suffix) throws IOException, InterruptedException {
         try {
             return new FilePath(this,act(new SecureFileCallable<String>() {
                 private static final long serialVersionUID = 1L;
                 public String invoke(File dir, VirtualChannel channel) throws IOException {
-                    File f = File.createTempFile(prefix, suffix, dir);
-                    f.delete();
-                    f.mkdir();
-                    return f.getName();
+                    Path tempPath = Files.createTempDirectory(dir.toPath(), prefix, null);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/FilePath.java]: `dir.toPath()` may throw `InvalidPathException`. Since it is an API method, it has to be handled and wrapped to `IOException`

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
See the comment. I would suggest creating an Utility `static Path toPath(File file) throws IOException` method, which wraps `InvalidPathException` to `IOException`so that it will be checked.

**Comment by larrysingleton** [comment]
Suggested changes have been added.

File: core/src/main/java/hudson/FilePath.java (hunk 1 above)
Comment (larrysingleton) [reply core/src/main/java/hudson/FilePath.java]: Agreed - change added

File: core/src/main/java/hudson/FilePath.java [hunk 2]
```diff
@@ -2371,7 +2402,7 @@ public String validateAntFileMask(final String fileMasks, final int bound) throw
      * Default bound for {@link #validateAntFileMask(String, int, boolean)}.
      * @since 1.592
      */
-    public static int VALIDATE_ANT_FILE_MASK_BOUND = SystemProperties.getInteger(FilePath.class.getName() + ".VALIDATE_ANT_FILE_MASK_BOUND", 10000);
+    public static final int VALIDATE_ANT_FILE_MASK_BOUND = SystemProperties.getInteger(FilePath.class.getName() + ".VALIDATE_ANT_FILE_MASK_BOUND", 10000);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/FilePath.java]: This change is unrelated && it causes merge conflict.
Note that some fields are bob-final for a reason, so that they cannot be modified in System scripts

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/FilePath.java (hunk 2 above)
Comment (larrysingleton) [reply core/src/main/java/hudson/FilePath.java]: reverted

**Comment by oleg-nenashev** [comment]
Resolved the merge conflict

**Comment by larrysingleton** [comment]
@oleg-nenashev - is there anything else you need from me here?
The test case failure does not appear to be related.
Let me know otherwise.

**Comment by dwnusbaum** [comment]
Along with resolving the conflicts it might make sense to get rid of the newly added `FilePath#toPath` and use `Util#fileToPath` now that #3135 is merged.

**Review by oleg-nenashev — APPROVED** [review]

**Review by dwnusbaum — APPROVED** [review]
I have a minor comment but I think the PR is good as-is.

File: core/src/main/java/hudson/FilePath.java [hunk 3]
```diff
@@ -1400,17 +1403,25 @@ public String invoke(File dir, VirtualChannel channel) throws IOException {
...
                 public String invoke(File dir, VirtualChannel channel) throws IOException {
-                    File f = File.createTempFile(prefix, suffix, dir);
-                    f.delete();
-                    f.mkdir();
-                    return f.getName();
+                    Path tempPath = Files.createTempDirectory(Util.fileToPath(dir), name,  new FileAttribute<?>[] {});
+                    if (tempPath.toFile() == null) {
+                        throw new IOException("Failed to obtain file from path " + dir + " on " + remote);
+                    }
+                    return tempPath.toFile().getName();
```
[the remark this answers is not in the document]
Comment (larrysingleton) [reply core/src/main/java/hudson/FilePath.java]: tempPath.getFileName() is not null safe either - a check would still be necessary
https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#getFileName()

Happy to make the change either way - please advise.

[the remark this answers is not in the document]
Comment (dwnusbaum) [reply core/src/main/java/hudson/FilePath.java]: No, I think the current code is good. My thought was that we had enough information here to know that the path is non-empty and so the call would not return null, but maybe that's not the case.

File: core/src/main/java/hudson/FilePath.java [hunk 4]
```diff
@@ -1400,17 +1403,25 @@ public String invoke(File dir, VirtualChannel channel) throws IOException {
...
+            }
+            String name = StringUtils.join(s, ".");
             return new FilePath(this,act(new SecureFileCallable<String>() {
                 private static final long serialVersionUID = 1L;
                 public String invoke(File dir, VirtualChannel channel) throws IOException {
-                    File f = File.createTempFile(prefix, suffix, dir);
-                    f.delete();
-                    f.mkdir();
-                    return f.getName();
+                    Path tempPath = Files.createTempDirectory(Util.fileToPath(dir), name,  new FileAttribute<?>[] {});
```
Comment (dwnusbaum) [inline core/src/main/java/hudson/FilePath.java]: Found a potentially relevant issue while investigating [JENKINS-48407](https://issues.jenkins-ci.org/browse/JENKINS-48407). The old io code would create directories with permissions `0777 & (~umask)`, whereas the nio code will create directories with permissions `0700 & (~umask)`. It's easy to retain the old behavior by passing `PosixFileAttributes.asFileAttributes(EnumSet.allOf(PosixFilePermission.class))` as the file attribute parameter to `Files#createTemporaryDirectory` after checking `Functions#isWindows`. CC @oleg-nenashev do you think it is important to maintain the old permissions here?

**Comment by dwnusbaum** [comment]
Also I just noticed that `hudson.model.Util#createTempDir` appears to have the same issue, do you want to update it in this PR as well?

File: core/src/main/java/hudson/FilePath.java (hunk 4 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/FilePath.java]: Better safe than sorry. I would vote for retaining the permissions in this case for sure

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Needs a fix of the default permissions change (see the comment from @dwnusbaum )

**Comment by larrysingleton** [comment]
Add posix check as requested

**Comment by larrysingleton** [comment]
@oleg-nenashev / @dwnusbaum 
Requested changes have been added, build passes checks with no conflicts

File: core/src/main/java/hudson/FilePath.java [hunk 5]
```diff
@@ -77,16 +77,11 @@
...
-import java.nio.file.InvalidPathException;
-import java.nio.file.StandardCopyOption;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Comparator;
-import java.util.Enumeration;
-import java.util.List;
-import java.util.Map;
-import java.util.StringTokenizer;
+import java.nio.file.*;
```
Comment (dwnusbaum) [inline core/src/main/java/hudson/FilePath.java]: nit: I prefer to avoid `.*` imports to make it clear exactly what classes are being used.

File: core/src/main/java/hudson/FilePath.java [hunk 6]
```diff
@@ -77,16 +77,11 @@
...
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.*;
```
Comment (dwnusbaum) [inline core/src/main/java/hudson/FilePath.java]: Same

**Review by dwnusbaum — APPROVED** [review]
This looks good to me, although I would prefer to make the `.*` imports explicit again.

**Comment by larrysingleton** [comment]
@dwnusbaum - Updated to use explicit imports

**Review by dwnusbaum — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Fixed the merge conflict

File: core/src/test/java/hudson/FilePathTest.java [hunk 7]
```diff
@@ -641,7 +644,7 @@ public void testValidateCaseSensitivity() throws Exception {
             when(con.getInputStream())
               .thenReturn(someZippedContent());
 
-            assertTrue(d.installIfNecessaryFrom(url, null, null));
+            assertTrue(d.installIfNecessaryFrom(url, null, ""));
```
Comment (oleg-nenashev) [inline core/src/test/java/hudson/FilePathTest.java]: Does not seem to be related

Comment (oleg-nenashev) [reply core/src/test/java/hudson/FilePathTest.java]: The annotation has been introduced by @jglick in https://github.com/jenkinsci/jenkins/blame/09bcc5d6538b3cfffbf71228ebd1679e3e20d8b2/core/src/main/java/hudson/FilePath.java#L766 3 years ago, Does not matter now IMHO, I am fine with fixing it in this PR

**Review by oleg-nenashev — APPROVED** [review]
LGTM

File: core/src/test/java/hudson/FilePathTest.java (hunk 7 above)
Comment (larrysingleton) [reply core/src/test/java/hudson/FilePathTest.java]: True, it is not directly related - however the test case fails without the change.

Comment (oleg-nenashev) [reply core/src/test/java/hudson/FilePathTest.java]: @daniel-beck @jglick @dwnusbaum do we care about that? Maybe we need to do some extra NPE checks
